library/std/src/sync/mutex.rs

```diff
@@ -444,6 +511,14 @@ impl<'mutex, T: ?Sized> MutexGuard<'mutex, T> {
     unsafe fn new(lock: &'mutex Mutex<T>) -> LockResult<MutexGuard<'mutex, T>> {
         poison::map_result(lock.poison.borrow(), |guard| MutexGuard { lock, poison: guard })
     }
+
+    /// Immediately drops `self`, and consequently unlocks the mutex.
+    ///
+    /// This is equivalent to calling [`Drop::drop`] on the guard, but is more self-documenting.
+    #[unstable(feature = "guard_unlock", issue = "none")]
+    pub fn unlock(self) {
+        drop(self);
+    }
```

Comment:
`MutexGuard` doesn't have any methods right now, because it implements Deref. If the `T` already has an `.unlock()`, this will make it harder to call that `.unlock()` through this MutexGuard. Any existing code calling such an `.unlock()` would change meaning with this change.

Other types like `Box` solve this by not taking `self` in their methods (e.g. [`Box::into_raw`](https://doc.rust-lang.org/stable/alloc/boxed/struct.Box.html#method.into_raw)), and must be called with the more verbose `Box::into_raw(mybox)` syntax.

But I think `MutexGuard::unlock(guard);` wouldn't be a big improvement over `drop(guard);`. :(

What do you think?

Reply:
Hmm... I had not thought of that. I guess the primary difference in my mind was to make the code a bit self-documenting. It is not obvious to me when I see `drop(foo)` that a lock might be unlocked. I'm ok with the more verbose syntax, personally, but if we are going to do that perhaps we could do it like this instead:

```rust
impl Mutex {
  pub fn unlock(guard: MutexGuard) { drop(guard); }
}
```

Then, one would use it like

```rust
Mutex::unlock(guard);
```

which seems about as self-documenting as it gets...